File: 프로그래머스/lv2/12909. 올바른 괄호/올바른 괄호.py
from collections import deque
def solution(s):
    answer = True
    stack = deque()
    
    for ch in s:
        if ch == "(":
            stack.append(ch)
        elif not stack:
            answer = False
            break
        else:
            stack.pop()
    else:
        if stack:
            answer = False
    
    return answer

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
import java.util.ArrayDeque;
import java.util.Deque;


class Solution {
    boolean solution(String s) {
        boolean answer = true;

        Deque<Character> dq = new ArrayDeque<>();
        
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') {
                dq.offer('(');
                continue;
            }
            
            // 닫히는 괄호 일 때
            if (dq.isEmpty()) {
                answer = false;
                break;
            }
            dq.pollLast();
        }
        
        if(!dq.isEmpty()) answer = false;

        return answer;
    }
}